server/src/daemon/service.rs

```diff
@@ -0,0 +1,391 @@
+
+        let path = OwnedObjectPath::from(collection.path());
+        object_server.at(&path, collection).await?;
+
+        // perform prompt
+        self.manager.lock().unwrap().update_prompts_counter();
+        let prompt = Prompt::new(
+            Arc::clone(&self.manager),
+            self.manager.lock().unwrap().prompts_counter(),
+            None,
+        );
+        object_server
+            .at(prompt.path().to_owned(), prompt.to_owned())
+            .await?;
+
+        // signal
+        Self::collection_created(&ctxt, path.as_ref())
+            .await
+            .map_err::<ServiceError, _>(From::from)?;
+
+        Ok((path, prompt.path().to_owned()))
+    }
+
+    #[zbus(out_args("unlocked", "locked"))]
+    pub async fn search_items(
+        &self,
+        attributes: HashMap<&str, &str>,
+    ) -> Result<(Vec<Item>, Vec<Item>)> {
+        let items = self
+            .keyring
+            .search_items(&attributes)
+            .await
+            .map_err::<ServiceError, _>(From::from)?;
+
+        let mut unlocked: Vec<Item> = Vec::new();
+        let mut locked: Vec<Item> = Vec::new();
+
+        for item in items {
+            let attributes = item.attributes();
+            if attributes.get("locked").is_some() {
+                locked.push(item)
+            } else {
+                unlocked.push(item)
+            }
+        }
+
+        Ok((unlocked, locked))
+    }
+
+    #[zbus(out_args("unlocked", "prompt"))]
+    pub async fn unlock(
+        &mut self,
+        objects: Vec<OwnedObjectPath>,
+        #[zbus(signal_context)] ctxt: SignalContext<'_>,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(Vec<OwnedObjectPath>, ObjectPath)> {
+        // manages unlock state in memory
+        let mut unlocked: Vec<OwnedObjectPath> = Vec::new();
+
+        'main: for object in objects {
+            for collection in self.collections.read().await.iter() {
+                if collection.path() == *object {
+                    if collection.locked() {
+                        collection.set_locked(&ctxt, false).await;
+                        unlocked.push(object.clone());
+                    } else {
+                        break 'main;
+                    }
+                }
+            }
+        }
+
+        if unlocked.is_empty() {
+            unlocked.push(OwnedObjectPath::default());
+        }
+
+        // perform prompt
+        let lock: () = self.manager.lock().unwrap().update_prompts_counter();
+        let prompt = Prompt::new(
+            Arc::clone(&self.manager),
+            self.manager.lock().unwrap().prompts_counter(),
+            Some("u"),
+        );
+        object_server
+            .at(prompt.path().to_owned(), prompt.to_owned())
+            .await?;
+
+        Ok((unlocked, prompt.path().to_owned()))
+    }
+
+    #[zbus(out_args("locked", "prompt"))]
+    pub async fn lock(
+        &mut self,
```

Comment:
this is bad, don't use `&mut self` in any dbus function handler

Reply:
understood, thanks!